test/App/Fossa/API/BuildLinkSpec.hs

```diff
@@ -2,27 +2,34 @@
 
 module App.Fossa.API.BuildLinkSpec (spec) where
 
-import Test.Hspec
 import App.Fossa.API.BuildLink
-import Srclib.Types (Locator(Locator))
-import App.Fossa.FossaAPIV1 (Organization(Organization))
-import App.Types (ProjectRevision(ProjectRevision))
+import App.Fossa.FossaAPIV1 (Organization (Organization))
+import App.Types (ProjectRevision (ProjectRevision))
+import Control.Carrier.Diagnostics (DiagnosticsC, logDiagnostic)
+import Control.Monad (join)
+import Data.Functor.Identity (Identity (runIdentity))
 import Data.Text (Text)
+import Effect.Logger (IgnoreLoggerC, ignoreLogger)
 import Fossa.API.Types
+import Srclib.Types (Locator (Locator))
+import Test.Hspec
 import Text.URI.QQ
 
 simpleSamlPath :: Text
-simpleSamlPath = "https://app.fossa.com/account/saml/1?next=%2Fprojects%2Ffetcher123%252Bproject123%2Frefs%2Fbranch%2Fmaster123%2Frevision123"
+simpleSamlPath = "https://app.fossa.com/account/saml/1?next=/projects/fetcher123%2bproject123/refs/branch/master123/revision123"
 
 -- | Note the differences here between '%2F' and '%252F'.  The percent sign is re-encoded so that it's properly handled on the next redirect.
 gitSamlPath :: Text
-gitSamlPath = "https://app.fossa.com/account/saml/103?next=%2Fprojects%2Ffetcher%2540123%252Fabc%252Bgit%2540github.com%252Fuser%252Frepo%2Frefs%2Fbranch%2Fweird--branch%2Frevision%2540123%252Fabc"
+gitSamlPath = "https://app.fossa.com/account/saml/103?next=/projects/fetcher@123%252fabc%2bgit@github.com%252fuser%252frepo/refs/branch/weird--branch/revision@123%252fabc"
 
 fullSamlURL :: Text
-fullSamlURL = "https://app.fossa.com/account/saml/33?next=%2Fprojects%2Fa%252Bb%2Frefs%2Fbranch%2Fmaster%2Fc"
+fullSamlURL = "https://app.fossa.com/account/saml/33?next=/projects/a%2bb/refs/branch/master/c"
 
 simpleStandardURL :: Text
-simpleStandardURL = "https://app.fossa.com/projects/haskell%2B89%2Fspectrometer/refs/branch/master/revision123"
+simpleStandardURL = "https://app.fossa.com/projects/haskell+89%2fspectrometer/refs/branch/master/revision123"
+
+stripDiag :: DiagnosticsC (IgnoreLoggerC Maybe) a -> Maybe a
```

Comment:
I think we can build a better test helper that'd `expectationFailure` when diagnostics fail, which would give us better test error messages. similar to [how hspec-megaparsec works](https://hackage.haskell.org/package/hspec-megaparsec-2.2.0/docs/Test-Hspec-Megaparsec.html#v:parseSatisfies)

we could probably widely use it in other tests too. I think we do a lot of tedious diagnostics pattern matching elsewhere in our tests

Reply:
What would the type of that helper be?

Reply:
something like
```hs
idk :: DiagnosticsC IO a -> (a -> Expectation) -> Expectation

-- used like
test = do
  [...]
  idk (getBuildURLWithOrg org revision apiOpts locator) $ \url -> url `shouldBe` fullSamlURL
```

or maybe
```hs
idk :: Either FailureBundle (ResultBundle a) -> (a -> Expectation) -> Expectation

-- used like
test = do
  [...]
  let res = run . runDiagnostics $ getBuildURLWithOrg org revision apiOpts locator
  res `idk` \url -> url `shouldBe` fullSamlURL